# Natural Language Processing, Homework 1
 Dr. Sharon Yalov-Handzel

You can submit the homework as *.py or *.ipynb file.  
Answers to theoretical questions can be written in a separate WORD document.


Submitted by  
Michael Berger,  
Barak Simoni


  ## 1. Setup a python environment on your computer (Pycharm / google colab).


In [1]:
import pandas as pd
import spacy_stanza


  ## 2. Install the following libraries: pytorch, NLTK, spaCy
  


In [82]:
#conda install -y pytorch NLTK spaCy chardet pandasql

Retrieving notices: ...working... done
Channels:
 - defaults
 - conda-forge
Platform: linux-64
Solving environment: done


==> WARNING: A newer version of conda exists. <==
    current version: 24.7.1
    latest version: 24.11.2

Please update conda by running

    $ conda update -n base -c conda-forge conda



## Package Plan ##

  environment location: /media/michael/Data/anaconda3/envs/NLP

  added / updated specs:
    - pandasql


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    greenlet-3.1.1             |  py311h6a678d5_0         243 KB
    sqlalchemy-2.0.34          |  py311h00e1ef3_0         6.6 MB
    ------------------------------------------------------------
                                           Total:         6.8 MB

The following NEW packages will be INSTALLED:

  greenlet           pkgs/main/linux-64::greenlet-3.1.1-py311h6a678d5_0 
  pandasql           pkgs/main/linux


  ## 3. Write a program that loads the file spam.csv (source: KAGGLE)


In [83]:
import pandas as pd
import pandasql as ps 
import numpy as np
from IPython.display import Markdown, display
import re
import chardet

In [51]:
file = "data/spam.csv"
spam = pd.read_csv(file ,encoding='Windows-1252')

In [52]:

with open(file, 'rb') as rawdata:
    result = chardet.detect(rawdata.read(100000))
result

{'encoding': 'Windows-1252', 'confidence': 0.7270322499829184, 'language': ''}

In [53]:
spam

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,ham,"Go until jurong point, crazy.. Available only ...",NaN,NaN,NaN
1,ham,Ok lar... Joking wif u oni...,NaN,NaN,NaN
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,NaN,NaN,NaN
3,ham,U dun say so early hor... U c already then say...,NaN,NaN,NaN
4,ham,"Nah I don't think he goes to usf, he lives aro...",NaN,NaN,NaN
...,...,...,...,...,...
5567,spam,This is the 2nd time we have tried 2 contact u...,NaN,NaN,NaN
5568,ham,Will Ì_ b going to esplanade fr home?,NaN,NaN,NaN
5569,ham,"Pity, * was in mood for that. So...any other s...",NaN,NaN,NaN
5570,ham,The guy did some bitching but I acted like i'd...,NaN,NaN,NaN


In [54]:
spam[spam.iloc[:,2].notna()]

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
95,spam,Your free ringtone is waiting to be collected....,PO Box 5249,"MK17 92H. 450Ppw 16""",NaN
281,ham,\Wen u miss someone,the person is definitely special for u..... B...,why to miss them,"just Keep-in-touch\"" gdeve.."""
444,ham,\HEY HEY WERETHE MONKEESPEOPLE SAY WE MONKEYAR...,HOWU DOIN? FOUNDURSELF A JOBYET SAUSAGE?LOVE ...,NaN,NaN
671,spam,SMS. ac sun0819 posts HELLO:\You seem cool,"wanted to say hi. HI!!!\"" Stop? Send STOP to ...",NaN,NaN
710,ham,Height of Confidence: All the Aeronautics prof...,"this wont even start........ Datz confidence..""",NaN,NaN
899,spam,Your free ringtone is waiting to be collected....,PO Box 5249,"MK17 92H. 450Ppw 16""",NaN
1038,ham,"Edison has rightly said, \A fool can ask more ...",GN,GE,"GNT:-)"""
1127,ham,"Height of \Oh shit....!!\"" situation: A guy th...",".;-):-D""",NaN,NaN
1266,ham,\Hey sorry I didntgive ya a a bellearlier hunny,just been in bedbut mite go 2 thepub l8tr if u...,NaN,NaN
1384,ham,"Storming msg: Wen u lift d phne, u say \HELLO\...","bt not his girlfrnd... G o o d n i g h t . . .@""",NaN,NaN


Some messages got spread into several columns of the dataframe  
This happened due to inherent structure of messages, which contain all the possible characters,  
The presence of " and , chars in the messages makes separating tricky.  

We can solve it easily by concatenating the columns back into one column

In [56]:
def empty_join(lst, sep = ", "):
  return sep.join([x for x in lst if x])

In [57]:
spam.columns = "v1,v2,v3,v4,v5".split(",")
spam["msg"] = spam[['v2', 'v3', 'v4', 'v5']].fillna('').agg(empty_join, axis=1)
spam[["v1","msg"]]

,v1,msg
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."
...,...,...
5567,spam,This is the 2nd time we have tried 2 contact u...
5568,ham,Will Ì_ b going to esplanade fr home?
5569,ham,"Pity, * was in mood for that. So...any other s..."
5570,ham,The guy did some bitching but I acted like i'd...



  ## 4. Compute simple statistics on the data:   


In [58]:

rows = spam.msg.apply(func=lambda n: re.split("[\d\W]+", n.lower()))
words = sum(rows.values.tolist(),[])

wordsPerMess = rows.apply(func = len)

unique, counts = np.unique(words, return_counts=True)
wc = pd.DataFrame({"token":unique, "count": counts})

tokens = wc.sort_values("count",ascending=False)

In [60]:
show = rows[:20].apply(lambda r:  " ".join(r))
print("\n".join(show))

go until jurong point crazy available only in bugis n great world la e buffet cine there got amore wat 
ok lar joking wif u oni 
free entry in a wkly comp to win fa cup final tkts st may text fa to to receive entry question std txt rate t c s apply over s
u dun say so early hor u c already then say 
nah i don t think he goes to usf he lives around here though
freemsg hey there darling it s been week s now and no word back i d like some fun you up for it still tb ok xxx std chgs to send å to rcv
even my brother is not like to speak with me they treat me like aids patent 
as per your request melle melle oru minnaminunginte nurungu vettam has been set as your callertune for all callers press to copy your friends callertune
winner as a valued network customer you have been selected to receivea å prize reward to claim call claim code kl valid hours only 
had your mobile months or more u r entitled to update to the latest colour mobiles with camera for free call the mobile update co free on 

In [61]:
ans = f"""
number of SMS messages: {spam.shape[0]},  
number of spams: {sum(spam.v1=="spam")},   
total word count: {len(words)},   
averaged number of words per message: {wordsPerMess.mean():.2f},   
5 most frequent words in the file: {tokens.token.head(5).values},   
number of rare words (appear only once in the file): {sum(tokens["count"] == 1)}.  
"""

display(Markdown(ans))


number of SMS messages: 5572,  
number of spams: 747,   
total word count: 91723,   
averaged number of words per message: 16.46,   
5 most frequent words in the file: ['' 'i' 'to' 'you' 'a'],   
number of rare words (appear only once in the file): 3891.  



  ## 5. Perform tokenization twice: 


  ### a) by using NLTK library

  
Barak

first let's view the data:


In [3]:
spam.head()

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,ham,"Go until jurong point, crazy.. Available only ...",NaN,NaN,NaN
1,ham,Ok lar... Joking wif u oni...,NaN,NaN,NaN
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,NaN,NaN,NaN
3,ham,U dun say so early hor... U c already then say...,NaN,NaN,NaN
4,ham,"Nah I don't think he goes to usf, he lives aro...",NaN,NaN,NaN


It seems that v1 contains the classification of the message and v2 contains the message itself. so we would like to tokenize v2:

In [4]:
# Tokenization using NLTK
import nltk
from nltk.tokenize import word_tokenize

# Ensure the NLTK tokenizer required resources are downloaded
nltk.download('punkt')

# Tokenize the 'v2' column from the DataFrame
spam['tokens_nltk'] = spam['v2'].apply(word_tokenize)

# Display tokenized data
spam[['v2', 'tokens_nltk']].head()

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\barak\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


,v2,tokens_nltk
0,"Go until jurong point, crazy.. Available only ...","[Go, until, jurong, point, ,, crazy, .., Avail..."
1,Ok lar... Joking wif u oni...,"[Ok, lar, ..., Joking, wif, u, oni, ...]"
2,Free entry in 2 a wkly comp to win FA Cup fina...,"[Free, entry, in, 2, a, wkly, comp, to, win, F..."
3,U dun say so early hor... U c already then say...,"[U, dun, say, so, early, hor, ..., U, c, alrea..."
4,"Nah I don't think he goes to usf, he lives aro...","[Nah, I, do, n't, think, he, goes, to, usf, ,,..."


  ### b) and by Spacy library. 

Michael

In [62]:
#! python -m spacy download en_core_web_sm

In [ ]:
import spacy
nlp = spacy.load("en_core_web_sm")

spam['spacy'] = spam['msg'].apply(nlp)

Take result of SpaCy nlp pileline as plain list of str - for comparison

In [103]:
spam["spacy_tokens"] = spam.spacy.apply(lambda row: [r.text for r in row])
spam.spacy_tokens

0       [Go, until, jurong, point, ,, crazy, .., Avail...
1                [Ok, lar, ..., Joking, wif, u, oni, ...]
2       [Free, entry, in, 2, a, wkly, comp, to, win, F...
3       [U, dun, say, so, early, hor, ..., U, c, alrea...
4       [Nah, I, do, n't, think, he, goes, to, usf, ,,...
                              ...                        
5567    [This, is, the, 2nd, time, we, have, tried, 2,...
5568    [Will, Ì, _, b, going, to, esplanade, fr, home...
5569    [Pity, ,, *, was, in, mood, for, that, ., So, ...
5570    [The, guy, did, some, bitching, but, I, acted,...
5571                  [Rofl, ., Its, true, to, its, name]
Name: spacy_tokens, Length: 5572, dtype: object

### c) Write what is the difference between the results.

We will start by finding the different tokens between the 2 tokenizators

In [ ]:
spam['token_diff'] = spam.apply(lambda row: set(row['spacy_tokens']).symmetric_difference(set(row['nltk_tokens'])), axis=1)

In [208]:
df = spam[spam.token_diff!=set()]
df.shape

(1851, 13)

Out of 5571 messages, only 1851 (33%) got tokenized differently.  
Let's check out what these differences are.

We will output the tokenisations for these messages into 2 separate text files and visualize their difference using a standard OS visual diff tool

In [ ]:
df.spacy_tokens.apply(" ".join).to_csv("spacy_tokens.csv")

In [ ]:
df.nltk_tokens.apply(" ".join).to_csv("nltk_tokens.csv")

![](tokenisation_diff.png)

Upon examining the diff we can conclude that most of the diffences are in the treatment of the punctuation marks.  
Most notably in non-standard use of them due to slang, informal writing and mistakes, characteristic to SMS message communication    


We can investigate this further.  
Let's examine individual tokens that get tokenized differently

In [210]:

spam['spacy_only'] = spam.apply(lambda row: list(set(row['spacy_tokens']).difference(set(row['nltk_tokens' ]))), axis=1)
spam['nltk_only']  = spam.apply(lambda row: list(set(row['nltk_tokens' ]).difference(set(row['spacy_tokens']))), axis=1)


In [175]:
so = spam.explode('spacy_only')[['spacy_only']]

q = """
select spacy_only, count(spacy_only) as cnt
from so 
group by spacy_only
order by cnt desc
limit 50
"""

so = ps.sqldf(q).head(50)

,spacy_only
0,NaN
1,NaN
2,rate)T&C
2,question(std
3,NaN
...,...
5568,Ì
5568,_
5569,NaN
5570,NaN


In [ ]:
no = spam.explode('nltk_only')[['nltk_only']]

q = """
select nltk_only, count(nltk_only) as cnt
from no 
group by nltk_only
order by cnt desc
limit 50
"""

no = ps.sqldf(q).head(50)

,nltk_only
0,NaN
1,NaN
2,rate
2,std
2,&
...,...
5567,BT-national-rate
5568,Ì_
5569,NaN
5570,NaN


And so we acquire 2 lists of tokens most frequently being tokenized differently between the two libraries:

In [187]:
together = pd.concat([no,so], axis=1)
together

,nltk_only,cnt,spacy_only,cnt
0,:,345,,302
1,),244,nt,291
2,lt,239,lt;#&gt,214
3,gt,234,/,170
4,#,212,:),157
5,'',143,"""",145
6,-,114,-,135
7,Ì_,97,_,111
8,dont,96,Ì,95
9,&,78,do,90


Most of them really appear to be caused by punctuation 


  ## 6. Perform lemmatization
  by NLTK and
  barak


In [5]:
# Perform lemmatization using NLTK
from nltk.stem import WordNetLemmatizer

# Ensure the required resources for lemmatization are downloaded
nltk.download('wordnet')
nltk.download('omw-1.4')

# Initialize lemmatizer
lemmatizer = WordNetLemmatizer()

# Define a function to lemmatize tokens
def lemmatize_tokens(tokens):
    return [lemmatizer.lemmatize(token) for token in tokens]

# Apply lemmatization to the 'tokens' column
spam['lemmatized_tokens_nltk'] = spam['tokens_nltk'].apply(lemmatize_tokens)

# Display lemmatized data
spam[['v2', 'tokens_nltk', 'lemmatized_tokens_nltk']].head()

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\barak\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\barak\AppData\Roaming\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


,v2,tokens_nltk,lemmatized_tokens_nltk
0,"Go until jurong point, crazy.. Available only ...","[Go, until, jurong, point, ,, crazy, .., Avail...","[Go, until, jurong, point, ,, crazy, .., Avail..."
1,Ok lar... Joking wif u oni...,"[Ok, lar, ..., Joking, wif, u, oni, ...]","[Ok, lar, ..., Joking, wif, u, oni, ...]"
2,Free entry in 2 a wkly comp to win FA Cup fina...,"[Free, entry, in, 2, a, wkly, comp, to, win, F...","[Free, entry, in, 2, a, wkly, comp, to, win, F..."
3,U dun say so early hor... U c already then say...,"[U, dun, say, so, early, hor, ..., U, c, alrea...","[U, dun, say, so, early, hor, ..., U, c, alrea..."
4,"Nah I don't think he goes to usf, he lives aro...","[Nah, I, do, n't, think, he, goes, to, usf, ,,...","[Nah, I, do, n't, think, he, go, to, usf, ,, h..."


  by Spacy libraries. 

Michael

In [80]:
for w in spam.spacy[4]:
  print(w, " | ", w.text, " | " ,w.lemma_)


Nah  |  Nah  |  nah
I  |  I  |  I
do  |  do  |  do
n't  |  n't  |  not
think  |  think  |  think
he  |  he  |  he
goes  |  goes  |  go
to  |  to  |  to
usf  |  usf  |  usf
,  |  ,  |  ,
he  |  he  |  he
lives  |  lives  |  live
around  |  around  |  around
here  |  here  |  here
though  |  though  |  though


  Write what is the difference between the results.




  ## 7. Perform stemming 
  by NLTK twice on both results of (6). Write what is the difference between the results.

Barak

### a. stemming on lemmatization made with nltk

In [6]:
# Perform stemming on lemmatized tokens using NLTK's SnowballStemmer
from nltk.stem import SnowballStemmer

# Initialize the SnowballStemmer
stemmer = SnowballStemmer(language='english')

# Define a function to stem lemmatized tokens
def stem_tokens(tokens):
    return [stemmer.stem(token) for token in tokens]

# Apply stemming to the 'lemmatized_tokens_nltk' column
spam['stemmed_tokens_nltk'] = spam['lemmatized_tokens_nltk'].apply(stem_tokens)

# Display stemmed data
spam[['v2', 'lemmatized_tokens_nltk', 'stemmed_tokens_nltk']].head()

,v2,lemmatized_tokens_nltk,stemmed_tokens_nltk
0,"Go until jurong point, crazy.. Available only ...","[Go, until, jurong, point, ,, crazy, .., Avail...","[go, until, jurong, point, ,, crazi, .., avail..."
1,Ok lar... Joking wif u oni...,"[Ok, lar, ..., Joking, wif, u, oni, ...]","[ok, lar, ..., joke, wif, u, oni, ...]"
2,Free entry in 2 a wkly comp to win FA Cup fina...,"[Free, entry, in, 2, a, wkly, comp, to, win, F...","[free, entri, in, 2, a, wkli, comp, to, win, f..."
3,U dun say so early hor... U c already then say...,"[U, dun, say, so, early, hor, ..., U, c, alrea...","[u, dun, say, so, earli, hor, ..., u, c, alrea..."
4,"Nah I don't think he goes to usf, he lives aro...","[Nah, I, do, n't, think, he, go, to, usf, ,, h...","[nah, i, do, n't, think, he, go, to, usf, ,, h..."


### b. stemming on lemmatization made with spacy
TBD when Michael finishes his part


  ## 8. Identify a spam message where its removal from the spam.csv dataset would:




  ### a)Reduce the total number of stemmed tokens
  
BArak

We will find the spam message with the highest stemmed token count

In [8]:
# Identify spam messages in the dataset
spam_messages = spam[spam['v1'] == 'spam'].copy()

# Calculate the count of stemmed tokens for each spam message
spam_messages['stemmed_token_count_nltk'] = spam_messages['stemmed_tokens_nltk'].apply(len)

# Find the spam message with the highest token count
message_to_remove = spam_messages.loc[spam_messages['stemmed_token_count_nltk'].idxmax()]

# Display the identified message to remove and its token count
print("Spam message to remove:")
print(message_to_remove[['v2', 'stemmed_token_count_nltk']])

Spam message to remove:
v2                          URGENT! Your mobile number *************** WON...
stemmed_token_count_nltk                                                   46
Name: 2327, dtype: object



  ### b)Maintain the exact same number of lemmatized tokens
Michael

If no such message exists, provide a detailed explanation of why this scenario is
impossible or improbable.




  ## 9. Identify a spam message where its removal from the spam.csv dataset would:



  ### a)Reduce the total number of lemmatized tokens

Barak

We will apply the same logic that we used on the stemmed tokens, on the lemmatized tokens as we did to the stemmed ones:

In [9]:
# Identify spam messages in the dataset
spam_messages = spam[spam['v1'] == 'spam'].copy()

# Calculate the count of lemmatized tokens for each spam message
spam_messages['lemmatized_token_count_nltk'] = spam_messages['lemmatized_tokens_nltk'].apply(len)

# Find the spam message with the highest lemmatized token count
message_to_remove = spam_messages.loc[spam_messages['lemmatized_token_count_nltk'].idxmax()]

# Display the identified message to remove and its token count
print("Spam message to remove:")
print(message_to_remove[['v2', 'lemmatized_token_count_nltk']])


Spam message to remove:
v2                             URGENT! Your mobile number *************** WON...
lemmatized_token_count_nltk                                                   46
Name: 2327, dtype: object



  ### b)Maintain the exact same number of stemmed tokens

  Michael
If no such message exists, provide a detailed explanation of why this scenario is
impossible or improbable.




  ## 10. Compare the results of
  NLTK before and after tokenization, lemmatization and stemming

  BArak


Let's break it down to steps:
* First of all, before tokenization each message contained a full sentence. After tokenization - nltk seperated each sentence to words. We can also see that For example the word don't was separated to: "do", "n't".
* Let's use code to see the differences between tokenization and lemmatization:

In [14]:
# Compare 'tokens_nltk' and 'lemmatized_tokens_nltk' columns and print unique differences
unique_differences = set()  # Use a set to store unique differences

for index, row in spam.iterrows():
    original_tokens = row['tokens_nltk']
    lemmatized_tokens = row['lemmatized_tokens_nltk']

    # Check if there are any differences between the two lists
    if original_tokens != lemmatized_tokens:
        for original_token, lemmatized_token in zip(original_tokens, lemmatized_tokens):
            if original_token != lemmatized_token and (original_token, lemmatized_token) not in unique_differences:
                unique_differences.add((original_token, lemmatized_token))
                print(f"Token '{original_token}' was lemmatized as '{lemmatized_token}'.")

Token 'goes' was lemmatized as 'go'.
Token 'lives' was lemmatized as 'life'.
Token 'aids' was lemmatized as 'aid'.
Token 'has' was lemmatized as 'ha'.
Token 'as' was lemmatized as 'a'.
Token 'friends' was lemmatized as 'friend'.
Token 'hours' was lemmatized as 'hour'.
Token 'months' was lemmatized as 'month'.
Token 'mobiles' was lemmatized as 'mobile'.
Token 'chances' was lemmatized as 'chance'.
Token 'pounds' was lemmatized as 'pound'.
Token 'words' was lemmatized as 'word'.
Token 'times' was lemmatized as 'time'.
Token 'its' was lemmatized as 'it'.
Token 'sucks' was lemmatized as 'suck'.
Token 'knows' was lemmatized as 'know'.
Token 'does' was lemmatized as 'doe'.
Token 'us' was lemmatized as 'u'.
Token 'was' was lemmatized as 'wa'.
Token 'cheers' was lemmatized as 'cheer'.
Token 'timings' was lemmatized as 'timing'.
Token 'watts' was lemmatized as 'watt'.
Token 'sms' was lemmatized as 'sm'.
Token 'inches' was lemmatized as 'inch'.
Token 'calls' was lemmatized as 'call'.
Token 'messa

We can see that plural suffixes were constantly removed in the lemmatization process. However, they were sometimes unnecessarily removed (For example: the word "was" was turned into "wa" and it should have stayed "was". Now let's compare lemmatization and stemming:

In [16]:
# Compare 'lemmatized_tokens_nltk' and 'stemmed_tokens_nltk' columns and print unique differences
unique_differences_lemmatization_stemming = set()  # Use a set to store unique differences

for index, row in spam.iterrows():
    lemmatized_tokens = row['lemmatized_tokens_nltk']
    stemmed_tokens = row['stemmed_tokens_nltk']

    # Check if there are any differences between the two lists
    if lemmatized_tokens != stemmed_tokens:
        for lemmatized_token, stemmed_token in zip(lemmatized_tokens, stemmed_tokens):
            if lemmatized_token != stemmed_token and (lemmatized_token, stemmed_token) not in unique_differences_lemmatization_stemming:
                unique_differences_lemmatization_stemming.add((lemmatized_token, stemmed_token))
                print(f"Lemmatized token '{lemmatized_token}' was stemmed as '{stemmed_token}'.")

Lemmatized token 'Go' was stemmed as 'go'.
Lemmatized token 'crazy' was stemmed as 'crazi'.
Lemmatized token 'Available' was stemmed as 'avail'.
Lemmatized token 'only' was stemmed as 'onli'.
Lemmatized token 'bugis' was stemmed as 'bugi'.
Lemmatized token 'Cine' was stemmed as 'cine'.
Lemmatized token 'amore' was stemmed as 'amor'.
Lemmatized token 'Ok' was stemmed as 'ok'.
Lemmatized token 'Joking' was stemmed as 'joke'.
Lemmatized token 'Free' was stemmed as 'free'.
Lemmatized token 'entry' was stemmed as 'entri'.
Lemmatized token 'wkly' was stemmed as 'wkli'.
Lemmatized token 'FA' was stemmed as 'fa'.
Lemmatized token 'Cup' was stemmed as 'cup'.
Lemmatized token 'May' was stemmed as 'may'.
Lemmatized token 'Text' was stemmed as 'text'.
Lemmatized token 'receive' was stemmed as 'receiv'.
Lemmatized token 'T' was stemmed as 't'.
Lemmatized token 'C' was stemmed as 'c'.
Lemmatized token 'apply' was stemmed as 'appli'.
Lemmatized token 'U' was stemmed as 'u'.
Lemmatized token 'early' w

We notice the following differences:
* All words were lowercased.
* All suffixes (such as plural suffixes, ing, able, etc.) were removed.
* 'y' suffix was turned into 'i'.


  ## 11. Compare the results of
  Spacy before and after tokenization and lemmatization.

MIchael


  ## 12. Install the library BeautifulSoup



  ## 13. Choose any url page. 
  You can choose a page of yourself in a social media (e.g., facebook, Instagram, Linkdin, Twitter), or anything else



  ## 14. Perform data scrapping of this url. 
  You can learn how to do it from the following links:  
https://www.edureka.co/blog/web-scraping-with-python/  
https://zindi.africa/learn/a-beginners-guide-to-scraping-data-from-social-media  
or, ask CHATGPT to program it for you.  

Michael


  ## 15. Perform steps 5-11 on the text achieved in (14).


  ## 16. Download one of your whatsapp chats into *.txt file
  (at least 50 messages in Hebrew).

  Barack


In [19]:
# Read the _chat.txt file using pandas
chat = pd.read_csv('data/_chat.txt', sep='\t', encoding='utf-8')  # Adjust arguments like `sep` and `encoding` if needed
chat.head()

,"[24.7.2020, 18:58:28] משפוחה: ‏ההודעות והשיחות מוצפנות מקצה לקצה. לאף אחד מחוץ לצ'אט הזה, גם לא ל-WhatsApp, אין אפשרות לקרוא אותן ולהאזין להן."
0,"[25.7.2020, 13:18:31] אבא: אימא שבת שלום"
1,"[25.7.2020, 13:37:46] סבתא מזל: by"
2,"[2.8.2020, 9:19:21] ~ ברק ג'נח: חחח איזה חמודי..."
3,"[2.8.2020, 9:19:34] ~ ברק ג'נח: ואת הרבה יותר ..."
4,"[2.8.2020, 9:19:52] גילי כהן: תודה 🤩 אם קצת שי..."



  ## 17. Repeat steps 5-11 on the text file achieved in (16)


In [22]:
## Step 5 - Tokenization using NLTK (for Hebrew text)


from nltk.tokenize import word_tokenize
from nltk.stem import SnowballStemmer
from nltk.stem import WordNetLemmatizer
import nltk

# Ensure required NLTK data is downloaded
nltk.download('punkt')

chat['tokens_nltk'] = chat.iloc[:, 0].apply(word_tokenize)

# Display tokenized data
chat[['tokens_nltk']].head()

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\barak\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


,tokens_nltk
0,"[[, 25.7.2020, ,, 13:18:31, ], אבא, :, אימא, ש..."
1,"[[, 25.7.2020, ,, 13:37:46, ], סבתא, מזל, :, by]"
2,"[[, 2.8.2020, ,, 9:19:21, ], ~, ברק, ג'נח, :, ..."
3,"[[, 2.8.2020, ,, 9:19:34, ], ~, ברק, ג'נח, :, ..."
4,"[[, 2.8.2020, ,, 9:19:52, ], גילי, כהן, :, תוד..."


In [30]:
# still requires fixing !!!!!!
# ## Step 5 - Tokenization using spaCy (for Hebrew text)
# import stanza
# import spacy_stanza
#
# # Download and set up Stanza for Hebrew language
# stanza.download("he")  # Hebrew
# nlp_stanza = spacy_stanza.load_pipeline("he")
#
#
# # Tokenize the chat text using spaCy with Stanza for Hebrew processing
# chat['tokens_spacy'] = chat.iloc[:, 0].apply(lambda text: [token.text for token in nlp_stanza(text)])
#
# # Display tokenized data
# chat[['tokens_spacy']].head()


2024-12-23 05:23:59 INFO: Downloading default packages for language: he (Hebrew) ...


KeyboardInterrupt: 

In [31]:
## Step 6 - Lemmatization using NLTK (for Hebrew text)

from nltk.stem import WordNetLemmatizer
nltk.download('wordnet')

stemmer = SnowballStemmer("english")
lemmatizer = WordNetLemmatizer()
chat['lemmatized_tokens_nltk'] = chat['tokens_nltk'].apply(lambda tokens: [lemmatizer.lemmatize(token) for token in tokens])

# Display lemmatized data
chat[['lemmatized_tokens_nltk']].head()

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\barak\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


,lemmatized_tokens_nltk
0,"[[, 25.7.2020, ,, 13:18:31, ], אבא, :, אימא, ש..."
1,"[[, 25.7.2020, ,, 13:37:46, ], סבתא, מזל, :, by]"
2,"[[, 2.8.2020, ,, 9:19:21, ], ~, ברק, ג'נח, :, ..."
3,"[[, 2.8.2020, ,, 9:19:34, ], ~, ברק, ג'נח, :, ..."
4,"[[, 2.8.2020, ,, 9:19:52, ], גילי, כהן, :, תוד..."


In [ ]:
# should be fixed after we fix the spacy part !
# ## Step 6 - Lemmatization using spaCy (for Hebrew text)
#
# if nlp:
#     chat['lemmatized_tokens_spacy'] = chat.iloc[:, 0].apply(lambda text: [token.lemma_ for token in nlp(text)])
# else:
#     chat['lemmatized_tokens_spacy'] = None
#
# # Display lemmatized data
# chat[['lemmatized_tokens_spacy']].head()

In [34]:
from nltk.stem.isri import ISRIStemmer

# Initialize NLTK's ISRIStemmer for Hebrew-like stemming
stemmer = ISRIStemmer()

# Apply stemming for each token in the lemmatized tokens
chat['stemmed_tokens_nltk'] = chat['lemmatized_tokens_nltk'].apply(
    lambda tokens: [stemmer.stem(token) for token in tokens]
)

# Display stemmed data
chat[['stemmed_tokens_nltk']].head()

,stemmed_tokens_nltk
0,"[[, 25.7.2020, ,, 13:18:31, ], אבא, :, אימא, ש..."
1,"[[, 25.7.2020, ,, 13:37:46, ], סבתא, מזל, :, by]"
2,"[[, 2.8.2020, ,, 9:19:21, ], ~, ברק, ג'נח, :, ..."
3,"[[, 2.8.2020, ,, 9:19:34, ], ~, ברק, ג'נח, :, ..."
4,"[[, 2.8.2020, ,, 9:19:52, ], גילי, כהן, :, תוד..."


In [ ]:
# again - should be fixed like all spacy parts
# ## Step 7 - Stemming on SpaCy Lemmatization Results (for Hebrew text)
# if chat['lemmatized_tokens_spacy'].notnull().all():
#     chat['stemmed_tokens_spacy'] = chat['lemmatized_tokens_spacy'].apply(lambda tokens: [stemmer.stem(token) for token in tokens])
# else:
#     chat['stemmed_tokens_spacy'] = None
#
# # Display stemmed data
# chat[['stemmed_tokens_spacy']].head()

In [37]:
# Step 8 - Identify a chat message where its removal from the dataset would:

# Import libraries
from collections import Counter

# a) Reduce the total number of stemmed tokens
chat['stemmed_token_count_nltk'] = chat['stemmed_tokens_nltk'].apply(len)  # Count stemmed tokens
message_to_remove_a = chat.loc[chat['stemmed_token_count_nltk'].idxmax()]  # Find message with the highest count

print("Chat message to remove (to reduce total stemmed tokens):")
print(message_to_remove_a[['tokens_nltk', 'stemmed_tokens_nltk', 'stemmed_token_count_nltk']])

# b) Maintain the exact same number of lemmatized tokens
chat['lemmatized_token_count_nltk'] = chat['lemmatized_tokens_nltk'].apply(len)  # Count lemmatized tokens

# Identify a message where removing it doesn’t change the total lemmatized token count
lemmatized_counter = Counter([token for tokens in chat['lemmatized_tokens_nltk'] for token in tokens])
message_to_remove_b = None

for index, row in chat.iterrows():
    token_counts = Counter(row['lemmatized_tokens_nltk'])
    if all(lemmatized_counter[token] - count > 0 for token, count in token_counts.items()): # amount of tokens stayed the same
        message_to_remove_b = row
        break

if message_to_remove_b is not None:
    print("Chat message to remove (to maintain the same number of lemmatized tokens):")
    print(message_to_remove_b[['tokens_nltk', 'lemmatized_tokens_nltk', 'lemmatized_token_count_nltk']])
else:
    print("No chat message exists that would maintain the same number of lemmatized tokens after removal.")

Chat message to remove (to reduce total stemmed tokens):
tokens_nltk                 [[, 18.9.2020, ,, 10:42:03, ], שרונה, וינרב, :...
stemmed_tokens_nltk         [[, 18.9.2020, ,, 10:42:03, ], שרונה, וינרב, :...
stemmed_token_count_nltk                                                  106
Name: 322, dtype: object
Chat message to remove (to maintain the same number of lemmatized tokens):
tokens_nltk                    [המון, מזל, טוב, ברק, גילי, ריף, ולירוי]
lemmatized_tokens_nltk         [המון, מזל, טוב, ברק, גילי, ריף, ולירוי]
lemmatized_token_count_nltk                                           7
Name: 67, dtype: object


In [38]:
# Step 9 - Identify a chat message where its removal from the dataset would:

# a) Reduce the total number of lemmatized tokens
chat['lemmatized_token_count_nltk'] = chat['lemmatized_tokens_nltk'].apply(len)  # Count lemmatized tokens
message_to_remove_a = chat.loc[chat['lemmatized_token_count_nltk'].idxmax()]  # Find message with the highest count

print("Chat message to remove (to reduce total lemmatized tokens):")
print(message_to_remove_a[['tokens_nltk', 'lemmatized_tokens_nltk', 'lemmatized_token_count_nltk']])

# b) Maintain the exact same number of stemmed tokens
chat['stemmed_token_count_nltk'] = chat['stemmed_tokens_nltk'].apply(len)  # Count stemmed tokens

# Identify a message where removing it doesn’t change the total stemmed token count
stemmed_counter = Counter([token for tokens in chat['stemmed_tokens_nltk'] for token in tokens])
message_to_remove_b = None

for index, row in chat.iterrows():
    token_counts = Counter(row['stemmed_tokens_nltk'])
    if all(stemmed_counter[token] - count > 0 for token, count in token_counts.items()):  # Amount of tokens stayed the same
        message_to_remove_b = row
        break

if message_to_remove_b is not None:
    print("Chat message to remove (to maintain the same number of stemmed tokens):")
    print(message_to_remove_b[['tokens_nltk', 'stemmed_tokens_nltk', 'stemmed_token_count_nltk']])
else:
    print("No chat message exists that would maintain the same number of stemmed tokens after removal.")

Chat message to remove (to reduce total lemmatized tokens):
tokens_nltk                    [[, 18.9.2020, ,, 10:42:03, ], שרונה, וינרב, :...
lemmatized_tokens_nltk         [[, 18.9.2020, ,, 10:42:03, ], שרונה, וינרב, :...
lemmatized_token_count_nltk                                                  106
Name: 322, dtype: object
Chat message to remove (to maintain the same number of stemmed tokens):
tokens_nltk                 [המון, מזל, טוב, ברק, גילי, ריף, ולירוי]
stemmed_tokens_nltk         [המון, מזל, טוב, ברק, גילי, ריף, ולירוי]
stemmed_token_count_nltk                                           7
Name: 67, dtype: object


In [39]:
# Step 9 - Identify a chat message where its removal from the dataset would:

# a) Reduce the total number of lemmatized tokens
chat['lemmatized_token_count_nltk'] = chat['lemmatized_tokens_nltk'].apply(len)  # Count lemmatized tokens
message_to_remove_a = chat.loc[chat['lemmatized_token_count_nltk'].idxmax()]  # Find message with the highest count

print("Chat message to remove (to reduce total lemmatized tokens):")
print(message_to_remove_a[['tokens_nltk', 'lemmatized_tokens_nltk', 'lemmatized_token_count_nltk']])

# b) Maintain the exact same number of stemmed tokens
chat['stemmed_token_count_nltk'] = chat['stemmed_tokens_nltk'].apply(len)  # Count stemmed tokens

# Identify a message where removing it doesn’t change the total stemmed token count
stemmed_counter = Counter([token for tokens in chat['stemmed_tokens_nltk'] for token in tokens])
message_to_remove_b = None

for index, row in chat.iterrows():
    token_counts = Counter(row['stemmed_tokens_nltk'])
    if all(stemmed_counter[token] - count > 0 for token, count in token_counts.items()):  # Amount of tokens stayed the same
        message_to_remove_b = row
        break

if message_to_remove_b is not None:
    print("Chat message to remove (to maintain the same number of stemmed tokens):")
    print(message_to_remove_b[['tokens_nltk', 'stemmed_tokens_nltk', 'stemmed_token_count_nltk']])
else:
    print("No chat message exists that would maintain the same number of stemmed tokens after removal.")


Chat message to remove (to reduce total lemmatized tokens):
tokens_nltk                    [[, 18.9.2020, ,, 10:42:03, ], שרונה, וינרב, :...
lemmatized_tokens_nltk         [[, 18.9.2020, ,, 10:42:03, ], שרונה, וינרב, :...
lemmatized_token_count_nltk                                                  106
Name: 322, dtype: object
Chat message to remove (to maintain the same number of stemmed tokens):
tokens_nltk                 [המון, מזל, טוב, ברק, גילי, ריף, ולירוי]
stemmed_tokens_nltk         [המון, מזל, טוב, ברק, גילי, ריף, ולירוי]
stemmed_token_count_nltk                                           7
Name: 67, dtype: object


## Should also be performed with Spacy!

In [40]:
## Step 10 - Compare Tokenization, Lemmatization, and Stemming (Chat Data)

# Create dictionaries to store comparisons between steps
differences_tokenization = set()  # Differences between original text and tokenized
differences_lemmatization = set()  # Differences between tokenized and lemmatized
differences_stemming = set()  # Differences between lemmatized and stemmed

# Iterating through the rows of the chat dataset
for index, row in chat.iterrows():
    original_tokens = row['tokens_nltk']  # Tokenized message
    lemmatized_tokens = row['lemmatized_tokens_nltk']  # Lemmatized message
    stemmed_tokens = row['stemmed_tokens_nltk']  # Stemmed message

    # Compare tokenization with original tokens
    for token in original_tokens:
        if token not in row.iloc[0]:
            differences_tokenization.add(token)

    # Compare lemmatization with tokenized tokens
    for original_token, lemmatized_token in zip(original_tokens, lemmatized_tokens):
        if original_token != lemmatized_token:
            differences_lemmatization.add((original_token, lemmatized_token))

    # Compare stemming with lemmatized tokens
    for lemmatized_token, stemmed_token in zip(lemmatized_tokens, stemmed_tokens):
        if lemmatized_token != stemmed_token:
            differences_stemming.add((lemmatized_token, stemmed_token))

# Display unique differences for each step
print("Differences between Original and Tokenized (NLTK):")
for token in differences_tokenization:
    print(f"Original Token: {token}")

print("\nDifferences between Tokenized and Lemmatized (NLTK):")
for original_token, lemmatized_token in differences_lemmatization:
    print(f"Token '{original_token}' was lemmatized as '{lemmatized_token}'.")

print("\nDifferences between Lemmatized and Stemmed (NLTK):")
for lemmatized_token, stemmed_token in differences_stemming:
    print(f"Lemmatized token '{lemmatized_token}' was stemmed as '{stemmed_token}'.")

Differences between Original and Tokenized (NLTK):
Original Token: ''
Original Token: ``

Differences between Tokenized and Lemmatized (NLTK):
Token 'https' was lemmatized as 'http'.

Differences between Lemmatized and Stemmed (NLTK):
Lemmatized token 'مسرح' was stemmed as 'سرح'.
Lemmatized token 'المينا' was stemmed as 'ينا'.
